man/man3/fprintflux.3
.TH fprintflux 3  "LPJmL programmers manual"
.SH NAME
fprintflux, printflux \- Print global carbon and water fluxes
.SH SYNOPSIS
.nf
\fB#include "lpj.h"

void fprintflux(FILE *\fIfile\fB,Flux \fIflux\fB,Real cflux_total\fB,const Config *\fIconfig\fB);
void printflux(Flux \fIflux\fB,Real \fIcflux_total\fB,const Config *\fIconfig\fB);

.fi
.SH DESCRIPTION
Function prints global carbon and water fluxes. \fBprintflux()\fP writes output to standard output.
The arguments of \fBfprintflux\fP are:
.TP
.I file
File pointer to output stream.
.TP
.I flux
Global carbon and water fluxes. Must be calculated by \fBflux_sum()\fP.
.TP
.I cflux_total
Total flux of carbon. Must be calculated by \fBflux_sum()\fP.
.TP
.I config
LPJmL configuration data. Must be read by \fBreadconfig()\fP. 
.SH RETURN VALUE
None.
.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
flux_sum(3), readconfig(3)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

## [5.7.xx] - 2023-08-xx

### Removed

- removed large chunks of code in `allocation_daily_crop.c` that was inactive

### Fixed

- fixed output `PFT_VEGC` to not include `bm_inc.nitrogen` for crops in `daily_agriculture.c`

## [5.6.25] - 2023-07-29

### Added

- New outputs added for N fluxes from timber extraction and for N pools in wood product pools
- New outputs added for `estab_storage` C and N
- Added missing N fluxes, `estab_storage`, and reservoirs to `globalflux` output
- Added new fluxes to man pages

### Fixed 

- Corrected/added units in comments in output.h
- Corrected units in `flux_sum` man page
- Corrected `N2O_DENIT` output, which now no longer also includes `N2O_NIT` values
- Corrected writing of mineral N pools to outputs, so that also mineral N in frozen soil layers is included
- Corrected writing of `VEGN` for crops, so that `bm_inc.nitrogen` is no longer included to avoid double accounting

### Removed

- Removed unecessary initialization of output variables to zero in `update_annual.c`, which is done in `initoutputdata.c`


## [5.6.24] - 2023-07-29

### Fixed

-  The number of bands is now correctly checked for the `read_residue_data` input.


## [5.6.23] - 2023-07-28

### Changed

- Instead of the soil code the index of the grid cell is written into the NetCDF file for `"grid"` output.


## [5.6.22] - 2023-07-27

### Changed

- Configuration files `lpjml_netcdf.js`, `lpjml_non.js`, `param_non.js`, and `lpjparam_non.js` updated to latest lpjml version.

### Removed

- Predefined daily outputs (`"d_lai"`-`"d_pet"`) and corresponding functions removed.

- Version numbers removed from man pages.

### Fixed

- Variable `issocket` always initialized to `FALSE` in `readfilename.c`.


## [5.6.21] - 2023-07-21

### Changed

- Flag `"nitrogen_coupled"` disabled for lpjml offline runs. Water stress is always coupled with nitrogen stress. Only if lpjml is compiled with `-DCOUPLING_WITH_FMS` flag is enabled.

- Flag `"new_trf"` renamed to `"transp_suction_fcn"`.


## [5.6.20] - 2023-07-13

### Added

- New flags `"fix_co2"` and `"fix_co2_year"` added.

### Changed

- Flag `"istimber"` renamed to `"luc_timber"`.

- Flag `"new_phenology"` renamed to `"gsi_phenology"`.

- Flag `"shuffle_climate"` renamed to `"shuffle_spinup_climate"`.

- The settings for fixed input data have been updated. For climate and N deposition input an interval can be specified from which data is taken after the specified year. In can specified whether random shuffling or cycling is used:

```java
"fix_climate" : false,                /* fix climate after specified year */
"fix_climate_year" : 1901,            /* year after climate is fixed */
"fix_climate_interval" : [1901,1930],
"fix_climate_shuffle" : true,          /* randomly shuffle climate in the interval */
"fix_deposition_with_climate" : false, /* fix N deposition same as climate */
"fix_deposition" : false,              /* fix N deposition after specified year */
"fix_deposition_year" : 1901,          /* year after deposition is fixed */
"fix_deposition_interval" : [1901,1930],
"fix_deposition_shuffle" : true,       /* randomly shuffle depositions in the interval */
"fix_landuse" : false,                 /* fix land use after specfied year */
"fix_landuse_year" : 1901,             /* year after land use is fixed */
"fix_co2" : false,                     /* fix atmospheric CO2  after specfied year */
"fix_co2_year" : 1901,                 /* year after CO2 is fixed */
```

### Removed

- Obsolete flag `"const_climate"` removed.


## [5.6.19] - 2023-07-06

### Added

- Help option `-h` added to `bin2cdf` and `clm2cdf` utility.

- Target `lpjcheck` added to `Makefile` to create only the `lpjcheck` utility.

### Changed

- The version of all man pages has been set to the LPJmL version number, date of last modification removed.

- The different versions of the utilities have been replaced by a reference to the LPJmL version.

### Removed

- All html files have been removed from the repository.

- Obsolete link `lpj` to `lpjml` removed.


## [5.6.18] - 2023-07-04

### Fixed

- Soil array `wpwp` is now correctly initialized with `soilpar->wpwp` for the `"prescribed_soilpar"` setting.


## [5.6.17] - 2023-06-30

### Fixed

- Replaced logical `or` by logical `and` in `survive.c`. This solves issue [#309](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/309).


## [5.6.16] - 2023-06-30

### Fixed

- Removed excess `pft->nind` in all harvest routines in `harvest_stand.c`.


## [5.6.15] - 2023-06-30

### Added

- Check for matching size of discharge queues read from restart file added to `initriver()` function in `initdrain.c`. If the size differs from the value calculated from the river lengths the queues will be resized and initialized to zero. Error message is additionally printed.


## [5.6.14] - 2023-06-08

### Removed

- Support for the deprecated `*.conf` configuration file format has been removed. Only JSON files are allowed for configuration files and metafiles for input.

- Man pages for the description of the `*.conf` file format removed.

### Changed

- Strings in JSON configuration files can now be longer than 256 characters.


## [5.6.13] - 2023-05-16

### Added


src/lpj/fprintflux.c
/**************************************************************************************/
/**                                                                                \n**/
/**                f  p  r  i  n  t  f  l  u  x  .  c                              \n**/
/**                                                                                \n**/
/**     C implementation of LPJmL                                                  \n**/
/**                                                                                \n**/
/**     Function prints global carbon and water fluxes.                            \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define LINES_PER_HEADER 25

void fprintflux(FILE *file,          /**< Output file pointer */
                Flux flux,           /**< Carbon and water fluxes */
                Real cflux_total,    /**< Total carbon flux  (gC) */
                int year,            /**< Simulation year (AD) */
                const Config *config /**< LPJ configuration */
               )
{
  int i,tabs;
  Real convert;
  if(config->ngridcell > 2) convert = 1e-15;
    else convert = 1e-3;
  if((year<config->firstyear &&
     (year-config->firstyear+config->nspinup) % LINES_PER_HEADER==0) ||
     (year>=config->firstyear && (year-config->firstyear) % LINES_PER_HEADER==0))
  {
    /* print header */
    tabs=(config->fire) ? 5 : 4;
    if(config->withlanduse!=NO_LANDUSE)
      tabs+=2;
    fputs("\n       ",file);
    frepeatch(file,' ',(tabs*8-16)/2-1);
    fprintf(file,"Carbon flux (%ctC)",(convert==1e-15)  ? 'G' : 'k');
    frepeatch(file,' ',(tabs*8-16)/2);
    if(config->river_routing)
      frepeatch(file,' ',(config->withlanduse==NO_LANDUSE) ? 16 : 20);
    else
      frepeatch(file,' ',(config->withlanduse==NO_LANDUSE) ? 8 : 12);
    fprintf(file,"Water (%sm3)",(convert==1e-15)  ? "k" : "da");
    fprintf(file,"          CH4 fluxes (TgCH4)");
      fprintf(file,"              Carbon (%ctC)    ",(convert==1e-15)  ? 'T' : 'k');
    if(config->with_nitrogen)
    {
      if(config->river_routing)
        frepeatch(file,' ',(config->withlanduse==NO_LANDUSE) ? 4 : 8);
      fprintf(file,"                Nitrogen (%cgN)",(convert==1e-15)  ? 'T' : 'M');
    }
    fputs("\n       ",file);
    frepeatch(file,'-',tabs*8-1);
    if(config->withlanduse!=NO_LANDUSE)
      fputs(" ----------------------------------",file);
    else
      fputs(" --------------------------",file);
    if(config->river_routing)
      fputs("-----------",file);
    //fputs("--------", file);
    fputs(" -------------------------------- ----------------------------------------",file);
    if(config->with_nitrogen)
      fputs(" --------------------------------",file);
    fputc('\n',file);
    if(year<config->firstyear)
      fputs("Spinup ",file);
    else
      fputs("Year   ",file);
    fputs("NEP     estab  ",file);
    if(config->fire)
      fputs(" fire   ",file);
    if(config->withlanduse!=NO_LANDUSE)
      fputs(" harvest product",file);
    fputs(" NBP    ",file);
    fputs(" NPP    ",file);
    fputs(" transp     evap    interc ",file);
    if(config->withlanduse!=NO_LANDUSE)
      fputs(" wd     ",file);
    if(config->river_routing)
      fputs(" discharge ",file);
    fputs(" CH4 emiss.  CH4 sink   CH4 fire",file);
    if(config->withlanduse!=NO_LANDUSE)
      fputs("CH4 rice  ",file);
    fputs("SoilC     slowSoilC   VegC     Soil CH4  Soil NO3  Soil NH4  SoilN   ",file);
    if(config->with_nitrogen)
      fputs(" nuptake ndemand  nlosses ninflux",file);
    fputc('\n',file);
    fputs("------",file);
    for(i=0;i<tabs;i++)
      fputs(" -------",file);
    fputs(" ---------- ------- -------",file);
    if(config->withlanduse!=NO_LANDUSE)
      fputs(" -------",file);
    if(config->river_routing)
      fputs(" ----------",file);
    fputs(" ---------- ---------- ----------", file);
    if(config->withlanduse!=NO_LANDUSE)
      fputs(" ---------", file);
    fputs(" ----------  ---------- --------- --------- ---------- ---------", file);
    if(config->with_nitrogen)
      fputs(" ------- -------- ------- -------",file);
    fputc('\n',file);
  }
  /* print data */
  fprintf(file,"%6d %7.3f %7.3f",year,(flux.npp-flux.rh)*convert,flux.estab*convert);
  if(config->fire)
    fprintf(file," %7.3f",flux.fire*convert);
  if(config->withlanduse!=NO_LANDUSE)
    fprintf(file," %7.3f %7.3f",flux.harvest*convert,flux.product_turnover*convert);
  fprintf(file," %7.3f",cflux_total*convert);
  fprintf(file," %7.3f",flux.npp*convert);
  fprintf(file," %10.1f %7.1f %7.1f",
          flux.transp*convert*1000,flux.evap*convert*1000,flux.interc*convert*1000);
  if(config->withlanduse!=NO_LANDUSE)
    fprintf(file," %7.1f",flux.wd*convert*1000);
  if(config->river_routing)
    fprintf(file," %10.1f ",flux.discharge*convert);
  fprintf(file, " %9.3f ", flux.aCH4_emissions*convert*1000);
  fprintf(file, " %9.3f ", flux.aCH4_sink*convert*1000);
  fprintf(file, " %9.3f ", flux.aCH4_fire*convert*1000);
  if(config->withlanduse!=NO_LANDUSE)
    fprintf(file, " %9.3f ", flux.aCH4_rice*convert*1000);
  fprintf(file, " %9.3f ", flux.soilc*convert);
  fprintf(file, " %9.3f ", flux.soilc_slow*convert);
  fprintf(file, " %9.3f", flux.vegc*convert);
  fprintf(file, " %9.3f", flux.soil_CH4*convert);
  fprintf(file, " %9.3f", flux.soil_NO3*convert);
  fprintf(file, " %9.3f", flux.soil_NH4*convert);
  fprintf(file, " %9.3f ", flux.soiln*convert);
  if(config->with_nitrogen)
    fprintf(file," %7.1f %8.1f %7.1f %7.1f",flux.n_uptake*convert*1000,flux.n_demand*convert*1000,flux.n_outflux*convert*1000,
    flux.n_influx*convert*1000);
  fputc('\n',file);
} /* of 'fprintflux' */
